C/plugins/south/dht11/plugin.cpp

```diff
@@ -0,0 +1,110 @@
+/*
+ * FogLAMP south plugin.
+ *
+ * Copyright (c) 2018 OSisoft, LLC
+ *
+ * Released under the Apache 2.0 Licence
+ *
+ * Author: Mark Riddoch
+ */
+#include <dht11.h>
+#include <plugin_api.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <strings.h>
+#include <string>
+#include <logger.h>
+#include <plugin_exception.h>
+#include <config_category.h>
+#include <rapidjson/document.h>
+
+using namespace std;
+
+/**
+ * Default configuration
+ */
+#define CONFIG  "{\"plugin\" : { \"description\" : \"DHT11 plugin\", " \
+                        "\"type\" : \"string\", \"default\" : \"foglamp-dht11\" }, " \
+                "\"asset\" : { \"description\" : \"Asset name\", "\
+                        "\"type\" : \"string\", \"default\" : \"dht11\" }, " \
+                "\"pin\" : { \"description\" : \"RPi pin to which DHT11 is attached\", " \
+                        "\"type\" : \"integer\", \"default\" : \"7\" } "\
+                "}"
+
+
+/**
+ * The DHT11 plugin interface
+ */
+extern "C" {
+
+/**
+ * The plugin information structure
+ */
+static PLUGIN_INFORMATION info = {
+	"DHT11",                  // Name
+	"1.0.0",                  // Version
+	0,    			  // Flags
+	PLUGIN_TYPE_SOUTH,        // Type
+	"1.0.0",                  // Interface version
+	CONFIG                    // Default configuration
+};
+
+/**
+ * Return the information about this plugin
+ */
+PLUGIN_INFORMATION *plugin_info()
+{
+	return &info;
+}
+
+/**
+ * Initialise the plugin, called to get the plugin handle
+ */
+PLUGIN_HANDLE plugin_init(ConfigCategory *config)
+{
+	unsigned int pin=7;
+
+	if (config->itemExists("pin"))
+        {
+                pin = stoul(config->getValue("pin"), nullptr, 0);
+        }
+	
+	//Logger::getLogger()->info("DHT11 C++ plugin_init: config='%s' ", config->toJSON().c_str());
+
+	DHT11 *dht11= new DHT11(pin);
+	return (PLUGIN_HANDLE)dht11;
+}
+
+/**
+ * Start the Async handling for the plugin
+ */
+void plugin_start(PLUGIN_HANDLE *handle)
+{
+}
+
+/**
+ * Poll for a plugin reading
+ */
+Reading plugin_poll(PLUGIN_HANDLE *handle)
+{
+	DHT11 *dht11 = (DHT11*)handle;
+	return dht11->takeReading();
+}
+
+/**
+ * Reconfigure the plugin
+ */
+void plugin_reconfigure(PLUGIN_HANDLE *handle, string& newConfig)
+{
```

Comment:
This entry point is to allow for the dynamic reconfiguration of the plugin. The only things that matter are probably the asset, which might or might not need attention once that is implemented. Also the pion could be changed and currently that will be ignored.

Reply:
To start using a new GPIO pin to get data from the sensor, it is probably better to shutdown Pi or atleast the sensor and then plug it back in using the new GPIO pin. It would be better to power reset the sensor before changing the GPIO pin it is connected to. So dynamic (runtime) configuration change is probably not required. So even if I don't add logic to the reconfigure function, I think pin configuration change would still be stored into DB and would be used the next time dht11 plugin is started.

Should I add code to handle "asset name" change?

Reply:
Sure, hardware changes will probably result in a restart, but they may not, for example if two are plugged in and you want to swap from one to the other for some reason. Also changes to the asset code could easily happen at runtime.

Reply:
plugin_reconfigure is not being called from ManagementApi::configChange( ) at the moment. So I'll address this reconfigure part in a separate ticket, if that's ok with you.